### PR TITLE
Postgres Output Processor and WA Create Database Command

### PR DESCRIPTION
**Pull request by sdpenguin**
Add the WA Command to create a Postgres database, the schema for the database which stores the results found in the wa_output folder, and last but not least the Postgres Output Processor used to upload to the database.

### CODE REVIEW

File: wa/commands/create.py [hunk 1]
```diff
@@ -35,6 +40,138 @@
...
+                'Please supply a password for the account used to connect.')
+        # Create the database (override if exists)
+        if args.dbname == 'postgres':
+            raise ValueError('Databasename to create cannot be postgres.')
+        database_exists = False
+        try:
+            test_conn = connect(dbname=args.dbname, user=args.username,
+                                password=args.password, host=args.host, port=args.port)
+            test_conn.close()
+            database_exists = True
```
Comment (setrofim) [inline wa/commands/create.py]: Rather than using this flag, just perform error logging inside the `except` clause.

File: wa/commands/create.py [hunk 2]
```diff
@@ -35,6 +40,138 @@
...
+        except OperationalError:
+            pass
+
+        if database_exists and not args.force:
+            self.logger.error(
+                "Database {} already exists.".format(args.dbname) +
+                "Please specify the -f flag to create it afresh."
+            )
+            sys.exit()
```
Comment (setrofim) [inline wa/commands/create.py]: Raise `CommandError` rather than `sys.exit()`

File: wa/commands/create.py [hunk 3]
```diff
@@ -35,6 +40,138 @@
...
+
+        conn = connect(dbname='postgres', user=args.username,
```
Comment (setrofim) [inline wa/commands/create.py]: Break this block out into a function.

File: wa/commands/create.py [hunk 4]
```diff
@@ -35,6 +40,138 @@
...
+
+        conn.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
+        cursor = conn.cursor()
+        cursor.execute('DROP DATABASE IF EXISTS ' + args.dbname)
+        cursor.execute('CREATE DATABASE ' + args.dbname)
+        conn.commit()
+        cursor.close()
+        conn.close()
+
+        conn = connect(dbname=args.dbname, user=args.username,
```
Comment (setrofim) [inline wa/commands/create.py]: Break this block out into a function.

File: wa/commands/create.py [hunk 5]
```diff
@@ -35,6 +40,138 @@
...
+                                 help='The port on which the PostgreSQL server is running.')
+        self.parser.add_argument('-u', '--username', default='postgres',
+                                 help='The username with which to connect to the server.')
+        self.parser.add_argument('-p', '--password',
+                                 help='The password for the user account.')
+        self.parser.add_argument('-d', '--dbname', default='wa',
+                                 help='The name of the database to create.')
+        self.parser.add_argument('-f', '--force', action='store_true',
+                                 help='Force overwrite the existing database if one exists.')
+        self.parser.add_argument('-m', '--force-update-config', action='store_true',
```
Comment (setrofim) [inline wa/commands/create.py]: `-F` better than `-m`

File: wa/commands/create.py [hunk 6]
```diff
@@ -143,19 +282,23 @@ def initialize(self, context):
...
                 shutil.rmtree(package_path)
             else:
-                raise CommandError('Location "{}" already exists.'.format(package_path))
+                raise CommandError(
+                    'Location "{}" already exists.'.format(package_path))
         actual_package_path = os.path.join(package_path, name)
         os.makedirs(actual_package_path)
-        setup_text = render_template(setup_template_path, {'package_name': name, 'user': getpass.getuser()})
+        setup_text = render_template(
+            setup_template_path, {'package_name': name, 'user': getpass.getuser()})
```
Comment (setrofim) [inline wa/commands/create.py]: These sorts of changes that are nothing to do with the database should be in a separate commit.

File: wa/output_processors/postgresql.py [hunk 7]
```diff
@@ -0,0 +1,761 @@
...
+# limitations under the License.
+#
+
+
+import psycopg2
+import os
+import uuid
+import re
+import collections
+from psycopg2 import (connect,
```
Comment (setrofim) [inline wa/output_processors/postgresql.py]: - Space between stdlib and third-party dependencies
- import-guard these new dependencies

File: wa/output_processors/postgresql.py [hunk 8]
```diff
@@ -0,0 +1,761 @@
...
+    value = None
+
+    def __init__(self, level):
+        self.value = level
+
+    def return_original(self):
+        return self.value
+
+
+def adapt_ListOfX(adapt_X):
```
Comment (setrofim) [inline wa/output_processors/postgresql.py]: All of these need to be redone. "Adapt" is the wrong term here. You're serializing to postrgres format. Break this out into a separate module (or add to `wa/utils/serializer.py` -- though not sure if belongs there) and make sure serialization works propertly -- i.e. according to specification, not just for the inputs you've encountered. Ideally, see if there is already some library you can use for this -- e.g. is there anything in `psychopg2`?

File: wa/output_processors/postgresql.py [hunk 9]
```diff
@@ -0,0 +1,761 @@
...
+    def execute_sql_line_by_line(self, sql):
+        cursor = self.conn.cursor()
+        for line in sql.replace('\n', "").replace(";", ";\n").split("\n"):
+            if line and not line.startswith('--'):
+                cursor.execute(line)
+        cursor.close()
+        self.conn.commit()
+        self.conn.reset()
+
+    # def insert_large_object(self, large_object_path):
```
Comment (marcbonnici) [inline wa/output_processors/postgresql.py]: Commented out method definition left in.

File: wa/output_processors/postgresql.py [hunk 10]
```diff
@@ -0,0 +1,761 @@
...
+            augmentation_id = owner_id
+        elif parameter_type == "runtime":
+            job_uuid = self.current_job_uuid
+        else:
+            # boot parameters are not implemented
+            # device parameters are redundant due to the targets table
+            raise NotImplementedError
+        for parameter in parameter_dict:
+            self.current_parameter_uuid = uuid.uuid4()
+            from wa.utils.types import cpu_mask
```
Comment (marcbonnici) [inline wa/output_processors/postgresql.py]: Imports should all be at the beginning of the file.

File: wa/output_processors/postgresql.py [hunk 11]
```diff
@@ -0,0 +1,761 @@
...
+            if type(parameter_dict[parameter]) == cpu_mask:
+            self.cursor.execute(self.command_create_parameter,
```
Comment (marcbonnici) [inline wa/output_processors/postgresql.py]: Missing indent for `if` statement.

File: wa/output_processors/postgresql.py [hunk 12]
```diff
@@ -0,0 +1,761 @@
...
+        elif parameter_type == "runtime":
+            job_uuid = self.current_job_uuid
+        else:
+            # boot parameters are not implemented
+            # device parameters are redundant due to the targets table
+            raise NotImplementedError
+        for parameter in parameter_dict:
+            self.current_parameter_uuid = uuid.uuid4()
+            from wa.utils.types import cpu_mask
+            if type(parameter_dict[parameter]) == cpu_mask:
```
Comment (marcbonnici) [inline wa/output_processors/postgresql.py]: Maybe use `if isinstance(parameter_dict[parameter], cpu_mask):`?

File: wa/output_processors/postgresql.py [hunk 13]
```diff
@@ -0,0 +1,761 @@
...
+        elif parameter_type == "augmentation":
+            if job_specific:
+                job_uuid = self.current_job_uuid
+            augmentation_id = owner_id
+        elif parameter_type == "runtime":
+            job_uuid = self.current_job_uuid
+        else:
+            # boot parameters are not implemented
+            # device parameters are redundant due to the targets table
+            raise NotImplementedError
```
Comment (marcbonnici) [inline wa/output_processors/postgresql.py]: Should include parameters in Error message.

File: wa/output_processors/postgresql.py [hunk 14]
```diff
@@ -0,0 +1,761 @@
...
+                                    parameter_type
+                                )
+                                )
+
+    def connect_to_database(self):
+        dsn = "dbname={} user={} password={} host={} port={}".format(
+            self.dbname, self.username, self.password, self.host, self.port)
+        try:
+            self.conn = connect(dsn=dsn)
+        except Exception as e:
```
Comment (marcbonnici) [inline wa/output_processors/postgresql.py]: Do we know what sort of exceptions will be generated here rather than catching everything?

File: wa/output_processors/postgresql.py [hunk 15]
```diff
@@ -0,0 +1,761 @@
...
+        # (in order to store the information as soon as it becomes available)
+        if self.first_job_run:
+            self.target_uuid = uuid.uuid4()
+            # Run Artifacts initial population
+            for artifact in run_output.artifacts:
+                self.current_artifact_uuid = uuid.uuid4()
+                self.current_lobj = self.conn.lobject()
+                self.current_loid = self.current_lobj.oid
+                self.current_large_object_uuid = uuid.uuid4()
+                lobj_data = open(os.path.join(
```
Comment (marcbonnici) [inline wa/output_processors/postgresql.py]: Is this file ever closed?  Maybe use a `with` statement?

File: wa/output_processors/postgresql.py [hunk 16]
```diff
@@ -0,0 +1,761 @@
...
+                                    event.message
+                                )
+                                )
+        # Job artifacts
+        for artifact in output.artifacts:
+            self.current_artifact_uuid = uuid.uuid4()
+            self.current_lobj = self.conn.lobject()
+            self.current_loid = self.current_lobj.oid
+            self.current_large_object_uuid = uuid.uuid4()
+            lobj_data = open(os.path.join(
```
Comment (marcbonnici) [inline wa/output_processors/postgresql.py]: Is this file ever closed? Maybe use a with statement?

File: wa/output_processors/postgresql.py [hunk 17]
```diff
@@ -0,0 +1,761 @@
...
+                                )
+        # Run artifacts update
+        for artifact in output.artifacts:
+            if artifact in self.artifacts_already_added:
+                continue
+            self.current_artifact_uuid = uuid.uuid4()
+            self.current_lobj = self.conn.lobject()
+            self.current_loid = self.current_lobj.oid
+            self.current_large_object_uuid = uuid.uuid4()
+            lobj_data = open(os.path.join(
```
Comment (marcbonnici) [inline wa/output_processors/postgresql.py]: Is this file ever closed? Maybe use a with statement?

File: wa/commands/create.py [hunk 18]
```diff
@@ -35,6 +40,138 @@
...
+                        config['postgres'][item] = args_dict[item]
+            else:
+                config['postgres'] = OrderedDict([('host', args.host), ('port', args.port), (
+                    'dbname', args.dbname), ('username', args.username), ('password', args.password)])
+        else:
+            config['postgres'] = OrderedDict([('host', args.host), ('port', args.port), (
+                'dbname', args.dbname), ('username', args.username), ('password', args.password)])
+        config_file = open(args.config_file, 'w+')
+        yaml.dump(config, config_file)
+        config_file.close()
```
Comment (marcbonnici) [inline wa/commands/create.py]: Change this into a `with` statement.

File: wa/commands/create.py [hunk 19]
```diff
@@ -35,6 +40,138 @@
...
+
+        # Update config file
+        config_file = open(args.config_file, 'r')
+        config = yaml.load(config_file)
+        config_file.close()
+        if 'postgres' in config:
+            if not args.force_update_config:
+                self.logger.warning("The entry postgres already exists in the config file. " +
+                                    "Please specify the -m parameter to force an update.")
+            if type(config['postgres']) is OrderedDict:
```
Comment (marcbonnici) [inline wa/commands/create.py]: Maybe use `isinstance`?

File: wa/commands/create.py [hunk 20]
```diff
@@ -35,6 +40,138 @@
...
+            "Successfully created the database {}".format(args.dbname))
+
+        # Update config file
+        config_file = open(args.config_file, 'r')
+        config = yaml.load(config_file)
+        config_file.close()
+        if 'postgres' in config:
+            if not args.force_update_config:
+                self.logger.warning("The entry postgres already exists in the config file. " +
+                                    "Please specify the -m parameter to force an update.")
```
[the remark this answers is not in the document]
Comment (sdpenguin) [reply wa/commands/create.py]: Because the default is not to overwrite the config unless explicitly specified. This is for the user's benefit, to notify them of the fact that they can rerun the command with that flag to force update the config file too.

File: wa/output_processors/postgresql.py (hunk 8 above)
Comment (sdpenguin) [reply wa/output_processors/postgresql.py]: I don't think that adapt is the wrong term in this context, since it is the term used in the PostgreSQL documentation (as opposed to Cast, the inverse operation): http://initd.org/psycopg/docs/extensions.html#sql-adaptation-protocol-objects

I will be moving the adapters and the cast functions to a separate module called postgres_convert in utils which I will import in the output processor.

File: wa/commands/create.py [hunk 21]
```diff
@@ -35,6 +40,138 @@
...
+    def execute(self, state, args):
+        self.get_schema(self.schemafilepath)
+        # Display the version if needed and exit
+        if args.schema_version:
+            self.logger.info(
+                'The current version is {}'.format(self.schemaversion))
+            sys.exit()
+        if not args.password:
+            self.parser.error(
+                'Please supply a password for the account used to connect.')
```
[the remark this answers is not in the document]
Comment (sdpenguin) [reply wa/commands/create.py]: I've made the password parameter optional, and the database connection will be attempted without one if a password is not given. It will notify the user if the connection is rejected.

File: wa/commands/create.py [hunk 22]
```diff
@@ -35,6 +45,168 @@
...
+            '-r', '--config-file', default=settings.user_config_file,
+            help='Path to the config file to be updated.')
+        self.parser.add_argument(
+            '-x', '--schema-version', action='store_true',
+            help='Display the current schema version.')
+
+    def execute(self, state, args):  # pylint: disable=too-many-branches
+        # Upon psycopg2 import, the import guard checks if the import was successful
+        if not psycopg2:
+            raise CommandError('Psycopg2 could not be imported and is needed for wa create database: {}'.format(import_error_msg))
```
Comment (marcbonnici) [inline wa/commands/create.py]: You might want to break this line up.

File: wa/commands/create.py [hunk 23]
```diff
@@ -13,16 +13,26 @@
...
 
 from devlib.utils.types import identifier
+try:
+    import psycopg2
+except ImportError as e:
+    psycopg2 = None
+    import_error_msg = e.args[0] if e.args else str(e)
+else:
+    from psycopg2 import connect, OperationalError, extras
+    from psycopg2.extensions import ISOLATION_LEVEL_AUTOCOMMIT
```
Comment (marcbonnici) [inline wa/commands/create.py]: Rather than doing `try` except`, `else`, just try doing the imports directly.

File: wa/commands/create.py [hunk 24]
```diff
@@ -35,6 +45,168 @@
...
+        if not args.password:
+            try:
+                test_conn = connect(dbname='postgres', user=args.username,
+                                    host=args.postgres_host, port=args.postgres_port)
+                test_conn.close()
+            except OperationalError:
+                raise CommandError(
+                    'Passwordless connection to server failed.' +
+                    ' Change peer to trust in pg_hba.conf.' +
+                    ' Alternatively, supply a password as an argument or in .pgpass.')
```
[the remark this answers is not in the document]
Comment (sdpenguin) [reply wa/commands/create.py]: I think it does, since looking at the possible Exceptions raised by psycopg2 are available here: http://initd.org/psycopg/docs/module.html

File: wa/utils/postgres_convert.py [hunk 25]
```diff
@@ -0,0 +1,206 @@
...
+
+import re
+
+try:
+    from psycopg2 import InterfaceError
+    from psycopg2.extensions import AsIs
+except ImportError as e:
+    InterfaceError = None
+    AsIs = None
+    import_error_msg = e.args[0] if e.args else str(e)
```
[the remark this answers is not in the document]
Comment (sdpenguin) [reply wa/utils/postgres_convert.py]: Why not?

[the remark this answers is not in the document]
Comment (marcbonnici) [reply wa/utils/postgres_convert.py]: Sorry I meant the `import_error_msg` is not used here

File: wa/commands/create.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (marcbonnici) [reply wa/commands/create.py]: The Exceptions might be the same but wouldn't the exception message be different? 

Also is there anyway to distinguish what would happen if the server is setup for passwordless connections but fails for another reason that raises the same `OperationalError`?

File: wa/commands/postgres_schema.sql [hunk 26]
```diff
@@ -0,0 +1,170 @@
...
+DROP TABLE IF EXISTS ResourceGetters;
+DROP TABLE IF EXISTS Events;
+DROP TABLE IF EXISTS Targets;
+DROP TABLE IF EXISTS Jobs;
+DROP TABLE IF EXISTS Runs;
+
+DROP TYPE IF EXISTS status_enum;
+DROP TYPE IF EXISTS param_enum;
+
+CREATE TYPE status_enum AS ENUM ('UNKOWN(0)','NEW(1)','PENDING(2)','STARTED(3)','CONNECTED(4)', 'INITIALIZED(5)', 'RUNNING(6)', 'OK(7)', 'PARTIAL(8)', 'FAILED(9)', 'ABORTED(10)', 'SKIPPED(11)');
```
Comment (marcbonnici) [inline wa/commands/postgres_schema.sql]: UNKOWN -> UNKNOWN

File: setup.py [hunk 27]
```diff
@@ -41,7 +41,7 @@
     pass
 
 packages = []
-data_files = {}
+data_files = {'': [os.path.join(wa_dir, 'commands/postgres_schema.sql')]}
```
Comment (marcbonnici) [inline setup.py]: `commands/postgres_schema.sql` should be joined with `os.path.join`

File: wa/commands/create.py [hunk 28]
```diff
@@ -35,6 +44,171 @@
...
+                for item in ['postgres_host', 'postgres_port', 'dbname', 'username', 'password']:
+                    if item in config['postgres']:
+                        if args.force_update_config:
+                            config['postgres'][item] = args_dict[item]
+                    else:
+                        config['postgres'][item] = args_dict[item]
+            else:
+                if args.force_update_config:
+                    config['postgres'] = OrderedDict([('postgres_host', args.postgres_host), ('postgres_port', args.postgres_port), (
+                    'dbname', args.dbname), ('username', args.username), ('password', args.password)])
```
Comment (marcbonnici) [inline wa/commands/create.py]: PEP8 (E122 continuation line missing indentation or outdented) fix for this line needed.

File: wa/commands/create.py [hunk 29]
```diff
@@ -35,6 +44,171 @@
...
+                else:
+                    self.logger.warning("The entry postgres exists but is not in a correct json format. " +
+                        "Please fix this problem or specify the -F flag to force an update.")
```
Comment (marcbonnici) [inline wa/commands/create.py]: PEP8 fix (E128 continuation line under-indented for visual indent) needed for this line.

File: wa/commands/create.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (sdpenguin) [reply wa/commands/create.py]: Ok, I've used the predicate setup to check the various possible errors, and responded to them with meaningful information in the context.

File: wa/commands/create.py [hunk 30]
```diff
@@ -35,6 +44,171 @@
...
+                                        password=args.password, host=args.postgres_host, port=args.postgres_port)
+                else:
+                    test_conn = connect(dbname=args.dbname, user=args.username,
+                                        host=args.postgres_host, port=args.postgres_port)
+                test_conn.close()
+            except OperationalError:
+                raise CommandError(
+                    "Database {} already exists.".format(args.dbname) +
+                    "Please specify the -f flag to create it from afresh."
+                )
```
[the remark this answers is not in the document]
Comment (sdpenguin) [reply wa/commands/create.py]: Have found a way to check the exact Exception using the error message.

File: wa/commands/create.py [hunk 31]
```diff
@@ -35,6 +45,196 @@
...
+            for handle in possible_error_list:
+                predicate(e, handle)
+            raise e
+
+        # Update config file
+        with open(args.config_file, 'r') as config_file:
+            config = yaml.load(config_file)
+        if 'postgres' in config:
+            if not args.force_update_config:
+                self.logger.warning("The entry 'postgres' already exists in the config file. " +
```
Comment (setrofim) [inline wa/commands/create.py]: Raise `CommandError` here instead of logging a warning.

File: wa/output_processors/postgresql.py [hunk 32]
```diff
@@ -0,0 +1,601 @@
...
+#
+
+import os
+import uuid
+import collections
+
+try:
+    import psycopg2
+    from psycopg2 import (connect,
+                          extras)
```
Comment (setrofim) [inline wa/output_processors/postgresql.py]: Line break unnecessary here.

File: wa/output_processors/postgresql.py [hunk 33]
```diff
@@ -0,0 +1,601 @@
...
+from wa.framework.target.info import CpuInfo
+
+from devlib.target import KernelVersion, KernelConfig  # pylint: disable=wrong-import-order
+
+
+class PostgresqlResultProcessor(OutputProcessor):
+
+    name = 'postgres'
+    description = """
+    Stores results in an Postgresql database.
```
Comment (setrofim) [inline wa/output_processors/postgresql.py]: "a Postresql database"

File: wa/output_processors/postgresql.py [hunk 34]
```diff
@@ -0,0 +1,601 @@
...
+
+    Please see the documentation for information about the schema.
```
Comment (setrofim) [inline wa/output_processors/postgresql.py]: Link  to the relevant section of documentation here.

File: wa/output_processors/postgresql.py [hunk 35]
```diff
@@ -0,0 +1,601 @@
...
+
+    This can also be easily ascertained from the postgres_schema.sql file.
```
Comment (setrofim) [inline wa/output_processors/postgresql.py]: What file? How is the user supposed to know where to look for it? Provide absolute path (constructed from the location of this file.)

File: wa/output_processors/postgresql.py [hunk 36]
```diff
@@ -0,0 +1,601 @@
...
+    """
+    parameters = [
+        Parameter('username', default='postgres',
+                  global_alias='postgres_username',
```
Comment (setrofim) [inline wa/output_processors/postgresql.py]: Remove global aliases --  they're not appropriate here.

File: wa/output_processors/postgresql.py [hunk 37]
```diff
@@ -0,0 +1,601 @@
...
+                        self.run_uuid,
+                        resource_getter))
+                self.insert_parameters(
+                    'resource_getter',
+                    run_output.run_config.resource_getters[resource_getter],
+                    owner_id=self.current_resource_getter_uuid,
+                    job_specific=False)
+            self.first_job_run = False
+
+        # See wa/framework/output.py
```
Comment (setrofim) [inline wa/output_processors/postgresql.py]: Why?

File: wa/output_processors/postgresql.py [hunk 38]
```diff
@@ -0,0 +1,601 @@
...
+        # Job events
```
Comment (setrofim) [inline wa/output_processors/postgresql.py]: Each of these headings should be a separate method instead.

File: wa/output_processors/postgresql.py [hunk 39]
```diff
@@ -0,0 +1,601 @@
...
+                output.info.project,
+                retry_on_status,
+                output.run_config.max_retries,
+                output.run_config.bail_on_init_failure,
+                output.run_config.allow_phone_home,
+                output.info.uuid,
+                output.info.start_time,
+                output.info.end_time,
+                output.metadata,
+                run_uuid))
```
Comment (setrofim) [inline wa/output_processors/postgresql.py]: A lot of these should be added during `intialize` -- data should be uploaded as soon as possible.

File: wa/output_processors/postgresql.py [hunk 40]
```diff
@@ -0,0 +1,601 @@
...
+        # Run events
```
Comment (setrofim) [inline wa/output_processors/postgresql.py]: Again, the fact that you feel the need to add heading to blocks of code should be a hint that these ought to be broken out into their own methods.

File: wa/utils/postgres_convert.py [hunk 41]
```diff
@@ -0,0 +1,205 @@
...
+        return None
+
+    m = re.match(r"([^\()]*)\((\d*)\)", value)
+    name = str(m.group(1))
+    number = int(m.group(2))
+
+    if m:
+        return level(name, number)
+    else:
+        raise InterfaceError("Bad level representation: {}".format(value))
```
Comment (setrofim) [inline wa/utils/postgres_convert.py]: What happens if `InterfaceError` is `None` ?

File: wa/utils/postgres_convert.py [hunk 42]
```diff
@@ -0,0 +1,205 @@
...
+            master_sql_string = ""
+        for x in range(num_items):
+            # Remove trailing comma
+            master_list[x] = master_list[x].strip(",")
+            master_list[x] = "{" + master_list[x] + "}"
+            master_sql_string = master_sql_string + master_list[x] + ","
+        master_sql_string = master_sql_string.strip(",")
+        if num_items > 1:
+            master_sql_string = master_sql_string + "}"
+        return AsIs("'{}'".format(master_sql_string))
```
Comment (setrofim) [inline wa/utils/postgres_convert.py]: What happens if `AsIs` is `None`?

File: wa/output_processors/postgresql.py (hunk 39 above)
Comment (sdpenguin) [reply wa/output_processors/postgresql.py]: I have made sure to add all of the pieces of data as soon as they become available. Many of these are initially added in the export_job stages, but updated in the export_run stage.

File: wa/utils/postgres_convert.py (hunk 41 above)
Comment (sdpenguin) [reply wa/utils/postgres_convert.py]: The thing is that this function would never be defined and called on its own. It is of a format which adheres to the PostgreSQL casting API, so it would only ever be called by a user indirectly through the psycopg2.extensions.register_type function. In any case, I have now changed it so that if InterfaceError is None, it raises an ImportError anyway. The same applies to AsIs.

File: wa/output_processors/postgresql.py (hunk 35 above)
Comment (sdpenguin) [reply wa/output_processors/postgresql.py]: Done

File: wa/output_processors/postgresql.py (hunk 36 above)
Comment (sdpenguin) [reply wa/output_processors/postgresql.py]: Done

File: wa/output_processors/postgresql.py (hunk 37 above)
Comment (sdpenguin) [reply wa/output_processors/postgresql.py]: Removed - was for reference and an oversight was made in committing it.

File: wa/output_processors/postgresql.py (hunk 38 above)
Comment (sdpenguin) [reply wa/output_processors/postgresql.py]: Done

File: wa/output_processors/postgresql.py (hunk 40 above)
Comment (sdpenguin) [reply wa/output_processors/postgresql.py]: Done

File: wa/utils/postgres_convert.py (hunk 42 above)
Comment (sdpenguin) [reply wa/utils/postgres_convert.py]: See above. Done.

File: wa/output_processors/postgresql.py [hunk 43]
```diff
@@ -0,0 +1,478 @@
...
+
+
+class PostgresqlResultProcessor(OutputProcessor):
+
+    name = 'postgres'
+    description = """
+    Stores results in a Postgresql database.
+
+    This can also be easily ascertained from the postgres_schema.sql file:
+    ../commands/postgres_schema.sql
```
Comment (marcbonnici) [inline wa/output_processors/postgresql.py]: Don't need `also` here anymore, and I think the aim was to provide the full path of the file rather than the relative path as not all users will know the location of the installation. This can then be dynamically populated when using the show command etc.

Comment (sdpenguin) [reply wa/output_processors/postgresql.py]: Done

File: wa/output_processors/postgresql.py [hunk 44]
```diff
@@ -0,0 +1,600 @@
...
+    from psycopg2 import Error as Psycopg2Error
+except ImportError as e:
+    psycopg2 = None
+    import_error_msg = e.args[0] if e.args else str(e)
+
+from wa.utils import postgres_convert
+from wa import OutputProcessor, Parameter, OutputProcessorError
+from wa.utils.types import level
+from wa.framework.target.info import CpuInfo
+from devlib.target import KernelVersion, KernelConfig  # pylint: disable=wrong-import-order
```
[the remark this answers is not in the document]
Comment (setrofim) [reply wa/output_processors/postgresql.py]: `devlib` imports should go above WA imports.

File: wa/output_processors/postgresql.py [hunk 45]
```diff
@@ -0,0 +1,484 @@
...
+        self.run_uuid = uuid.UUID(str(uuid.uuid4()))
+        self.cursor.execute(self.sql_command['create_run'], (self.run_uuid,))
+        # Commit cursor commands
+        self.conn.commit()
+
+    def export_job_output(self, output, target_info, run_output):   # pylint: disable=too-many-branches, too-many-statements, too-many-locals
+        self.current_job_uuid = uuid.uuid4()
+        retry_on_status = postgres_convert.ListOfLevel(run_output.run_config.retry_on_status)
+        # Update the run each time
+        self.cursor.execute(
```
Comment (setrofim) [inline wa/output_processors/postgresql.py]: All of this stuff does not need to be done each time -- info and config stuff doesn't change; Also, those things should be written to the database earlier -- during run initialization, before the first job executes.

File: wa/output_processors/postgresql.py (hunk 39 above)
Comment (setrofim) [reply wa/output_processors/postgresql.py]: Again, there i s no need to write info and config stuff here -- only update things that (may) have changed.

File: wa/utils/postgres_convert.py (hunk 41 above)
Comment (setrofim) [reply wa/utils/postgres_convert.py]: The function *is* defined "on its own". If it's not intended to be called directly outside of this module, indicate this by prefixing its name with a single underscore.

Comment (setrofim) [reply wa/utils/postgres_convert.py]: Yes -- if it's exposed externally, you can't be sure that it will always be used correctly, and should guard against wrong usage.

**Comment by sdpenguin** [comment]
The pull request has been updated to address the remaining comments.

File: wa/utils/postgres_convert.py (hunk 41 above)
Comment (sdpenguin) [reply wa/utils/postgres_convert.py]: Done

**Comment by sdpenguin** [comment]
Pull request has been updated to add a space between the Devlib and WA imports.

**Comment by sdpenguin** [comment]
The pull request has been updated to address the remaining comments.

**Comment by sdpenguin** [comment]
Pull request has been updated following discussion in person.

**Comment by sdpenguin** [comment]
Pull request has been updated to address Marc's final 2 comments.

File: wa/output_processors/postgresql.py [hunk 46]
```diff
@@ -0,0 +1,482 @@
...
+
+try:
+    import psycopg2
+    from psycopg2 import (connect, extras)
+    from psycopg2 import Error as Psycopg2Error
+except ImportError as e:
+    psycopg2 = None
+    import_error_msg = e.args[0] if e.args else str(e)
+
+from devlib.target import KernelVersion, KernelConfig  # pylint: disable=wrong-import-order
```
[the remark this answers is not in the document]
Comment (sdpenguin) [reply wa/output_processors/postgresql.py]: I've removed it and grouped Devlib with the other 3rd party imports.
